Mathlib/Data/List/Sort.lean

```diff
@@ -0,0 +1,464 @@
+  | [] => sorted_nil
+  | a :: l => (sorted_insertionSort l).orderedInsert a _
+#align list.sorted_insertion_sort List.sorted_insertionSort
+
+end TotalAndTransitive
+
+end Correctness
+
+end InsertionSort
+
+/-! ### Merge sort -/
+
+
+section MergeSort
+
+-- TODO(Jeremy): observation: if instead we write (a :: (split l).1, b :: (split l).2), the
+-- equation compiler can't prove the third equation
+/-- Split `l` into two lists of approximately equal length.
+
+     split [1, 2, 3, 4, 5] = ([1, 3, 5], [2, 4]) -/
+@[simp]
+def split : List α → List α × List α
+  | [] => ([], [])
+  | a :: l =>
+    let (l₁, l₂) := split l
+    (a :: l₂, l₁)
+#align list.split List.split
+
+theorem split_cons_of_eq (a : α) {l l₁ l₂ : List α} (h : split l = (l₁, l₂)) :
+    split (a :: l) = (a :: l₂, l₁) := by rw [split, h]
+#align list.split_cons_of_eq List.split_cons_of_eq
+
+theorem length_split_le :
+    ∀ {l l₁ l₂ : List α}, split l = (l₁, l₂) → length l₁ ≤ length l ∧ length l₂ ≤ length l
+  | [], _, _, rfl => ⟨Nat.le_refl 0, Nat.le_refl 0⟩
+  | a :: l, l₁', l₂', h => by
+    cases' e : split l with l₁ l₂
+    injection (split_cons_of_eq _ e).symm.trans h; substs l₁' l₂'
+    cases' length_split_le e with h₁ h₂
+    exact ⟨Nat.succ_le_succ h₂, Nat.le_succ_of_le h₁⟩
+#align list.length_split_le List.length_split_le
+
+theorem length_split_lt {a b} {l l₁ l₂ : List α} (h : split (a :: b :: l) = (l₁, l₂)) :
+    length l₁ < length (a :: b :: l) ∧ length l₂ < length (a :: b :: l) := by
+  cases' e : split l with l₁' l₂'
+  injection (split_cons_of_eq _ (split_cons_of_eq _ e)).symm.trans h; substs l₁ l₂
+  cases' length_split_le e with h₁ h₂
+  exact ⟨Nat.succ_le_succ (Nat.succ_le_succ h₁), Nat.succ_le_succ (Nat.succ_le_succ h₂)⟩
+#align list.length_split_lt List.length_split_lt
+
+theorem perm_split : ∀ {l l₁ l₂ : List α}, split l = (l₁, l₂) → l ~ l₁ ++ l₂
+  | [], _, _, rfl => Perm.refl _
+  | a :: l, l₁', l₂', h => by
+    cases' e : split l with l₁ l₂
+    injection (split_cons_of_eq _ e).symm.trans h; substs l₁' l₂'
+    exact ((perm_split e).trans perm_append_comm).cons a
+#align list.perm_split List.perm_split
+
+/-- Merge two sorted lists into one in linear time.
+
+     merge [1, 2, 4, 5] [0, 1, 3, 4] = [0, 1, 1, 2, 3, 4, 4, 5] -/
+def merge : List α → List α → List α
+  | [], l' => l'
+  | l, [] => l
+  | a :: l, b :: l' => if a ≼ b then a :: merge l (b :: l') else b :: merge (a :: l) l'
+  termination_by merge l₁ l₂ => length l₁ + length l₂
+#align list.merge List.merge
+
+/-- Implementation of a merge sort algorithm to sort a list. -/
+def mergeSort : List α → List α
+  | [] => []
+  | [a] => [a]
+  | a :: b :: l => by
+    -- Porting note: rewrote to make `mergeSort_cons_cons` proof easier
+    let l₁ := (split (a :: b :: l)).1
+    let l₂ := (split (a :: b :: l)).2
+    have e : split (a :: b :: l) = ⟨l₁, l₂⟩ := rfl
+    have h := length_split_lt e
+    have := h.1
+    have := h.2
+    exact merge r (mergeSort l₁) (mergeSort l₂)
```

Comment:
I'm not sure this is the right solution - perhaps someone who knows about algorithms in lean4 could take a look. Maybe @digama0?